dumpling/dumpling.go

```diff
@@ -0,0 +1,246 @@
+		m.logger.Info("dump unit runs with injected error", zap.String("failpoint", "dumpUnitProcessWithError"), zap.Reflect("error", val))
+		msg, ok := val.(string)
+		if !ok {
+			msg = "unknown process error"
+		}
+		pr <- pb.ProcessResult{
+			IsCanceled: false,
+			Errors:     []*pb.ProcessError{unit.NewProcessError(errors.New(msg))},
+		}
+		failpoint.Return()
+	})
+
+	begin := time.Now()
+	errs := make([]*pb.ProcessError, 0, 1)
+	isCanceled := false
+
+	failpoint.Inject("dumpUnitProcessForever", func() {
+		m.logger.Info("dump unit runs forever", zap.String("failpoint", "dumpUnitProcessForever"))
+		<-ctx.Done()
+		failpoint.Return()
+	})
+
+	// NOTE: remove output dir before start dumping
+	// every time re-dump, loader should re-prepare
+	err := os.RemoveAll(m.cfg.Dir)
+	if err != nil {
+		m.logger.Error("fail to remove output directory", zap.String("directory", m.cfg.Dir), log.ShortError(err))
+	}
+
+	// Cmd cannot be reused, so we create a new cmd when begin processing
+	err = export.Dump(m.dumpConfig)
+
+	if err != nil {
+		dumplingExitWithErrorCounter.WithLabelValues(m.cfg.Name).Inc()
+		errs = append(errs, unit.NewProcessError(err))
+	} else {
+		select {
+		case <-ctx.Done():
+			isCanceled = true
+		default:
+		}
+	}
+
+	m.logger.Info("dump data finished", zap.Duration("cost time", time.Since(begin)))
+
+	pr <- pb.ProcessResult{
+		IsCanceled: isCanceled,
+		Errors:     errs,
+	}
+}
+
+// Close implements Unit.Close
+func (m *Dumpling) Close() {
+	if m.closed.Get() {
+		return
+	}
+	// do nothing, external will cancel the command (if running)
+	m.closed.Set(true)
+}
+
+// Pause implements Unit.Pause
+func (m *Dumpling) Pause() {
+	if m.closed.Get() {
+		m.logger.Warn("try to pause, but already closed")
+		return
+	}
+	// do nothing, external will cancel the command (if running)
+}
+
+// Resume implements Unit.Resume
+func (m *Dumpling) Resume(ctx context.Context, pr chan pb.ProcessResult) {
+	if m.closed.Get() {
+		m.logger.Warn("try to resume, but already closed")
+		return
+	}
+	// just call Process
+	m.Process(ctx, pr)
+}
+
+// Update implements Unit.Update
+func (m *Dumpling) Update(cfg *config.SubTaskConfig) error {
+	// not support update configuration now
+	return nil
+}
+
+// Status implements Unit.Status
+func (m *Dumpling) Status() interface{} {
+	// NOTE: try to add some status, like dumped file count
+	return &pb.DumpStatus{}
+}
+
+// Error implements Unit.Error
+func (m *Dumpling) Error() interface{} {
+	return &pb.DumpError{}
+}
+
+// Type implements Unit.Type
+func (m *Dumpling) Type() pb.UnitType {
+	return pb.UnitType_Dump
+}
+
+// IsFreshTask implements Unit.IsFreshTask
+func (m *Dumpling) IsFreshTask(ctx context.Context) (bool, error) {
+	return true, nil
+}
+
+// constructArgs constructs arguments for exec.Command
+func (m *Dumpling) constructArgs() (*export.Config, error) {
+	cfg := m.cfg
+	db := cfg.From
+
+	dumpConfig := export.DefaultConfig()
```

Comment:
1. Can we disable `StatusAddr` now? because we already have it in DM, and if we enable it, may we need more ports for the process.
2. Do we need to set `ServerInfo` based on `flavor`?

Reply:
1. I will block it.
2. dumpling will adjust `ServerInfo` during the `Dump` function.